internal/beater/java_attacher/os_utils.go

```diff
@@ -0,0 +1,142 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build !windows
+// +build !windows
+
+package javaattacher
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"syscall"
+)
+
+func (j *JavaAttacher) setRunAsUser(jvm *jvmDetails, cmd *exec.Cmd) error {
+	currentUser, err := user.Current()
+	if err != nil {
+		// If we cannot get the current user, then we assume apm-server
+		// is not the same as the target JVM user.
+		j.logger.Warnf("failed to get the current user: %v", err)
+	} else {
+		j.logger.Debugf("current user: %v", currentUser)
+	}
+
+	if currentUser.Gid == jvm.gid && currentUser.Uid == jvm.uid {
+		return nil // Users match, nothing to do.
+	}
+	uid, gid, err := parseUserIds(jvm.uid, jvm.gid)
+	if err != nil {
+		return err
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{Uid: uint32(uid), Gid: uint32(gid)},
+	}
+	return nil
+}
+
+// getAttacherJar finds an attacher jar based on the given uid.
+// In POSIX-compliant systems, it would be an attacher jar owned by the given user and with 0600 access mode.
+// If such was not yet created, this function calls createAttacherTempDir to create it and keep a mapping for it
+func (j *JavaAttacher) getAttacherJar(uid string) string {
+	j.tmpAttacherLock.Lock()
+	defer j.tmpAttacherLock.Unlock()
+	attacherJar, uidMapped := j.uidToAttacherJar[uid]
+	if uidMapped {
+		return attacherJar
+	}
+	attacherJar, err := j.createAttacherTempDir(uid)
+	if err != nil {
+		j.logger.Errorf("failed to create tmp dir for user %d, using the bundled attacher jar: %v", uid, err)
```

Comment:
Well, that is a matter of point of view.
The way I see it, whenever the APM Server is configured to try to attach an agent, the APM agent is the actual focus of the integration and the APM Server is an implementation detail. The user installed an Elastic agent on a machine that runs her Java application and want to get APM traces for it. If we can't do it, it is an error. The fact that the APM Server runs locally without errors is meaningless because it has no purpose other than sending data for this Java app.
And after the multiple back and forth, we made it clear that failing to create the temp dir is not something we want to try to recover from with a fallback

Reply:
>And after the multiple back and forth, we made it clear that failing to create the temp dir is not something we want to try to recover from with a fallback

I'm so confused. That's what I thought too, but you are trying to recover with a fallback here aren't you? You're logging an error and continuing...

Reply:
Sorry, you are right, the message is wrong.
There is no attempt to recover for this user, we are caching the fact that we cannot attach for JVMs of this user by the special case of empty jar path. We decided before that we do not want to fail the attacher entirely in this case, because the failure can be specific to user A and may not affect user B.

Reply:
Thanks for clarifying. I'm happy to leave it as an error log in that case, with the message amended.